### PR TITLE
Add TeamCity widget to dashboard

### PR DESCRIPTION
**Pull request by GSokol**
Widget shows several (default: 5) last build status-es + last changes author's username and date

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+FROM golang:1.8-alpine
```
Comment (ngalaiko) [inline Dockerfile:1]: nice

could you write template of dockerfile you would use to dockerize app for production?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+RUN mkdir -p /go/src/github.com/TheQuestionru/thequestion
```
Comment (ngalaiko) [inline Dockerfile:3]: you sure `GOPATH` will be set correctly?

File: makefile [hunk 3]
```diff
@@ -0,0 +1,5 @@
+all:
+	@ cd server \
+		&& govendor sync \
```
Comment (ngalaiko) [inline makefile:3]: it makes sense to run this *inside* container

File: server/schema/dashboard.go [hunk 4]
```diff
@@ -52,8 +52,11 @@ type DashboardRowView struct {
 type DashboardElementView struct {
 	*DashboardElement
 
-	Realtime types.NullInt64   `json:"realtime"`
-	Servers  []newrelic.Server `json:"servers,omitempty"`
+	Realtime          types.NullInt64   `json:"realtime"`
+	Servers           []newrelic.Server `json:"servers,omitempty"`
+	BuildStatuses     []string          `json:"lastBuilds,omitempty"`
+	LastChangesAuthor types.NullString  `json:"lastChangesAuthor,omitempty"`
+	LastChangesDate   types.NullString  `json:"lastChangesDate,omitempty"`
```
Comment (ngalaiko) [inline server/schema/dashboard.go:59]: why you use three fields instead of`TCBuildInfo` slice? 
it should not be done this way. 

buisness logic with this *last* things should be done somewhere else

File: server/schema/dashboard.go [hunk 5]
```diff
@@ -75,16 +78,22 @@ const (
...
 )
 
 func (t DashboardElementType) Clean() DashboardElementType {
-	switch t {
-	case DashboardElementTypeReportTemplate, DashboardElementTypeGARealtime,
-		DashboardElementTypeNRServers:
+	validValues := map[DashboardElementType]bool{
+		DashboardElementTypeReportTemplate: true,
+		DashboardElementTypeGARealtime:     true,
+		DashboardElementTypeNRServers:      true,
```
Comment (ngalaiko) [inline server/schema/dashboard.go:88]: just to follow our approach you should revert this changes.

i am practically sure that `switch` is more faster than `map`, but you can write a benchmark if not.

File: server/services/dashboard/dashboard.go [hunk 6]
```diff
@@ -88,6 +98,15 @@ func (t *dashboard) getDashboardGrid() ([][]*schema.DashboardElementView, error)
...
 					return nil, err
 				}
+			case schema.DashboardElementTypeTCBuilds:
+				tcBuildInfo, err := t.sideStats.TCBuildInfo()
+				if err != nil {
+					return nil, err
+				}
+
+				elementView.BuildStatuses = tcBuildInfo.BuildStatuses
+				elementView.LastChangesAuthor = types.NewNullString(tcBuildInfo.LastChangesAuthor)
```
Comment (ngalaiko) [inline server/services/dashboard/dashboard.go:108]: i think you should provide full structs to client, and it will take info it needs

File: server/services/stats/side/stats_side.go [hunk 7]
```diff
@@ -28,19 +29,31 @@ type Config struct {
 	Enabled              bool              `yaml:"Enabled"`
 	Schedule             string            `yaml:"Schedule"`
 	NewRelicApiKey       string            `yaml:"NewRelicApiKey"`
+	TeamCityAddress      string            `yaml:"TeamCityAddress"`
```
Comment (ngalaiko) [inline server/services/stats/side/stats_side.go:32]: why `user:pass` instead of `access_key`?

File: server/services/stats/side/tc_client.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	}
+}
+
+func NewTcClient(config Config) TcClient {
+	var authorizer teamcity.Authorizer
+	if config.TeamCityAddress == TestTeamCityAddress {
+		return NewTestTcClient()
+	}
+
+	if config.TeamCityLogin == "" && config.TeamCityPassword == "" {
```
Comment (ngalaiko) [inline server/services/stats/side/tc_client.go:47]: just throw panic here. we should have all credentials

File: server/services/stats/side/tc_client.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	GetTCBuildInfo() (*TCBuildInfo, error)
+}
+
+type tcClient struct {
+	client teamcity.Client
+	config Config
+}
+
+const (
+	tcBuildNumber = 5
```
Comment (ngalaiko) [inline server/services/stats/side/tc_client.go:30]: defaults shoud be in config

File: server/services/stats/side/tc_client.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	return nil
+}
+
+func (t *tcClient) setupChanges(result *TCBuildInfo) error {
+	changes, err := t.client.GetChanges(1)
+	if err != nil {
+		return err
+	}
+
+	if len(changes) == 0 {
```
Comment (ngalaiko) [inline server/services/stats/side/tc_client.go:88]: if there are no changes, return empty array or something else

why we need some constant error in dashboard if tc is not updated?

File: server/services/stats/side/test.go [hunk 11]
```diff
@@ -116,6 +119,20 @@ func (c *testGaClient) setQuestions(rows map[int64]*schema.AnalyticsGaRow) {
...
 
+type testTsClient struct{}
+
+func NewTestTcClient() TcClient {
+	return &testTsClient{}
+}
+
+func (t *testTsClient) GetTCBuildInfo() (*TCBuildInfo, error) {
+	return &TCBuildInfo{
+		BuildStatuses:     []string{"SUCCESS", "FAILURE", "ERROR"},
```
Comment (ngalaiko) [inline server/services/stats/side/test.go:130]: very strange struct

we have `slice` of build results and only author of last one.

but are you sure that slice sorted? and how know author of every else change?

File: Dockerfile (hunk 2 above)
Comment (GSokol) [reply Dockerfile:3]: What do you mean? `GOPATH` is defined in the parent package as `/go`

File: makefile (hunk 3 above)
Comment (GSokol) [reply makefile:3]: **No!**

1. It's better to make container as thin as it possible, so adding useless dependencies such as `git` and/or `govendor` is anti-pattern
2. You may have some private repositories in requirements list. If so, you are able to pass some access credentials to the container, which could cause the security leaks.

Comment (ngalaiko) [reply makefile:3]: you are right, but not is this case

if you are building binary file for production, you should have 2 container. one with all dependencies will build binary, and other one, `from alpline` will just run it

here, you run tests, so all deps must be inside docker, otherwise, why we need it? to run govendor, we need `go` and `govendor` on host machine. this dockerfile is like makefile

File: server/schema/dashboard.go (hunk 5 above)
Comment (GSokol) [reply server/schema/dashboard.go:88]: Of cause, `switch` is faster as well as takes less memory, but as I've notice in the commit message, I really believe that according to some metrics dashboard, which is actually not under big stress, readability is much more important that small efficiency reduce.

File: server/services/dashboard/dashboard.go (hunk 6 above)
Comment (GSokol) [reply server/services/dashboard/dashboard.go:108]: It's better to separate the view interface from interfaces of clients encapsulated in the view provider.
That's could help the view customers to be independent from the view provider realization and if it changes, that won't be matter for the customers.

File: server/schema/dashboard.go (hunk 5 above)
Comment (ngalaiko) [reply server/schema/dashboard.go:88]: it seems to me, `switch` is readable enough

File: server/services/stats/side/stats_side.go (hunk 7 above)
Comment (GSokol) [reply server/services/stats/side/stats_side.go:32]: I didn't want to spend much time, customizing third-party library for test task. I'm still don't

File: server/services/dashboard/dashboard.go (hunk 6 above)
Comment (ngalaiko) [reply server/services/dashboard/dashboard.go:108]: still, this info is pretty useless 

if previous build failed, you will never know author of it

File: server/services/stats/side/tc_client.go (hunk 8 above)
Comment (GSokol) [reply server/services/stats/side/tc_client.go:47]: And what's about the `GuestAuth`?

Comment (ngalaiko) [reply server/services/stats/side/tc_client.go:47]: `GuestAuth` is useless for some operations which require full access, which i think will be implemented later

File: server/services/dashboard/dashboard.go (hunk 6 above)
Comment (GSokol) [reply server/services/dashboard/dashboard.go:108]: Ok. The point I've done so was that I've asked @lisitsky what actually to show and he told me 

> Doesn't matter. For instance, show statuses of the last 5 builds and author and date of the last changes.

Comment (ngalaiko) [reply server/services/dashboard/dashboard.go:108]: it is really not a problem, but as for me, you could just grab build info

File: server/services/stats/side/test.go (hunk 11 above)
Comment (GSokol) [reply server/services/stats/side/test.go:130]: See [my previous comment](https://github.com/TheQuestionru/test-golang/pull/7#discussion_r125405012) ...

File: server/services/stats/side/tc_client.go (hunk 10 above)
Comment (GSokol) [reply server/services/stats/side/tc_client.go:88]: It may cause errors in view customers that expect the fields `LastChangesAuthor` and `LastChangesDate` are filled correctly.
Generally, of cause you are right, but it's hard to make the right decisions with such abstract task.

**Comment by GSokol** [comment]
Ok. If you want, I will provide the final CR-fix tomorrow.

**Comment by ngalaiko** [comment]
contact @lisitsky . he'll tell you
